File: boj/BinarySearch/BOJ_2805.cpp
/*
 백준 2805 나무 자르기
 21.07.06
 https://github.com/skyqnaqna/algorithm_study/
*/

#include <bits/stdc++.h>

#define INF 2147000000
#define endl "\n"

typedef long long ll;
typedef double dd;
typedef std::pair<int, int> pii;
typedef std::pair<ll, ll> pll;

using namespace std;

int n, m;
vector <ll> trees;

int main()
{
    ios::sync_with_stdio(0); cin.tie(0);

    cin >> n >> m;

    for (int i = 0; i < n; ++i) {
        int a; cin >> a;
        trees.push_back(a);
    }

    int mid, l = 0, r = 1000000000, ans = 0;

    while (l <= r) {
        mid = (l + r) / 2;


        ll sum = 0;
        for (auto &t : trees) {
            if (t > mid) {
                sum += t - mid;
            }
        }

        if (sum >= m) {
            l = mid + 1;
            ans = max(ans, mid);
        } else r = mid - 1;
    }

    cout << ans << endl;


    return 0;
}

File: boj/BinarySearch/BOJ_2805.swift
/*
 백준 2805 나무 자르기
 21.07.31
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let input = readLine()!.split(separator: " ").map{ Int(String($0))! }
let n = input[0], m = input[1]
let trees = readLine()!.split(separator: " ").map{ Int(String($0))! }

var l = 0, r = trees.max()!, mid = 0, ans = 0

while l <= r {
  mid = (l + r) / 2

  var sum = 0
  for tree in trees {
    if tree < mid { continue }
    sum += tree - mid
  }

  if sum >= m {
    ans = max(ans, mid)
    l = mid + 1
  } else {
    r = mid - 1
  }
}

print(ans)